community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/TxStateTransactionDataSnapshot.java

```diff
@@ -240,9 +240,15 @@ private void takeSnapshot( ReadableTxState state, StoreReadLayer storeReadLayer
 
     private Relationship relationship( long relId )
     {
-        state.relationshipVisit( relId, relationshipData );
-        return new RelationshipProxy( relationshipActions, relId,
-                relationshipData.startNode(), relationshipData.type(), relationshipData.endNode() );
+        if ( state.relationshipVisit( relId, relationshipData ) )
+        {
+            return new RelationshipProxy( relationshipActions, relId,
+                    relationshipData.startNode(), relationshipData.type(), relationshipData.endNode() );
+        }
+        else
+        {
+            return new RelationshipProxy( relationshipActions, relId );
+        }
```

Comment:
Would it perhaps be easier to write this as:

    RelationshipProxy result = new RelationshipProxy( relationshipActions, relId );
    state.relationshipVisit( relId, result );
    return result;

given that `RelationshipProxy` implements `RelationshipVisitor`. This would remove the need for the `RelationshipDataExtractor`.